Retrieve derived indices using a warming levels approach

nicolejkeeney opened the pull request:
# Description of PR
Retrieve derived indices using a warming levels approach. And, some unit tests. 

### Relevant motivation and context
I previously thought I had implemented this feature in PR #462 , but I forgot to enable the ability to retrieve the data using both a time-based and a warming level approach. 

### How to test 
Try to retrieve the Derived Indices (under the "Derived Indices" button in the GUI) using a Warming Level approach. Try with both the DataInterface GUI(i.e. `selections.show()`)  and with `get_data()`. 

The following are the derived variable options: 
- Fosberg fire weather index
- Effective Temperature
- NOAA Heat Index

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Definition of Done Checklist

#### Practical
- [x] 80% unit test coverage
- [x] Documentation
  - [x] All functions/adjusted functions documented in the [readthedocs](https://climakitae.readthedocs.io/en/latest/).
  - [x] Documentation is pushed
- [x] Complex code commented
- [x] Naming conventions followed
  - [x] Helper functions hidden with `_` before the name
- [x] Context of function is clearly provided
  - [x] Intent of function is provided
  - [x] How to test, so that it is not siloed on scientists and anyone can review
  - [x] Appropriate manual testing was completed
- [x] Any notebooks known to utilize the affected functions are still working
- [x] Linting completed and resolved

#### Conceptual
- [x] Doesn't replicate existing functionality
- [x] Aligns with general coding standard of existing functions
- [x] Matches desired functionality from users/scientists

vicford reviewed (changes requested):
`get_data` version:
- All variables were retrievable
- Noticed for effective temperature, if the timestep was hourly not daily it gives a non-specific error (see pic below)
![Screenshot 2024-11-25 at 1 00 28 PM](https://github.com/user-attachments/assets/1b4b1881-2e9f-4e5b-913d-d94bf9beba38)

GUI version:
- For derived index the warming level option disappears and defaults back to time

nicolejkeeney commented:
> `get_data` version:
> 
> * All variables were retrievable
> * Noticed for effective temperature, if the timestep was hourly not daily it gives a non-specific error (see pic below)

Effective temperature is only available for daily data, hence the error. The two other indices are only available for hourly. 
<img width="555" alt="Screenshot 2024-11-25 at 7 02 56 PM" src="https://github.com/user-attachments/assets/f5362db7-de42-488e-829a-73b8acbddfca">

That is a bad error message though... I'll think on how to return a more informative error message. Good find

nicolejkeeney commented:
> GUI version:
> 
> * For derived index the warming level option disappears and defaults back to time

Can you elaborate more on when this happens?

vicford commented:
> > GUI version:
> > 
> > * For derived index the warming level option disappears and defaults back to time
> Can you elaborate more on when this happens? Not sure I follow.

Okay when I went to take a quick video of what was happening, I couldn't replicate so was probably user error on my part!

vicford reviewed (approved):
Just an update to that bad error message would be good!

nicolejkeeney commented:
> Just an update to that bad error message would be good!

You actually found a bigger bug that was affecting more than just the derived indices. Basically anytime you put a bad input for `timescale`, that confusing error is raised (this can happen for other variables as well, since some variables are only available for hourly or only daily/monthly). 

Anyways good find. I just fixed it, I think. And added a relevant unit test to check that the function raises an appropriate error-- or, in this case, a print message.

claalmve commented on tests/data_interface/test_get_data.py:
```diff
@@ -0,0 +1,125 @@
+                )
+
+
+def TestAppropriateStringErrorReturnedIfBadInputGetData():
```
Is this supposed to be `class`?

claalmve commented on tests/data_interface/test_get_data.py:
```diff
@@ -0,0 +1,125 @@
+        # Error message we expect to be printed by the function
+        expected_print_message = "ERROR: No data found for your input values\n"
+
+        # NOTE: function PRINTS this message-- it does not return it as en error
```
nit: typo- "en error"

claalmve commented on tests/data_interface/test_get_data.py:
```diff
@@ -0,0 +1,125 @@
+        # Error message we expect to be printed by the function
+        expected_print_message = "ERROR: Function argument warming_level requires a float/int or list of floats/ints input. Your input: <class 'str'> \nReturning None\n"
+
+        # NOTE: function PRINTS this message-- it does not return it as en error
```
nit: typo- "en error"

claalmve commented:
One consideration for the test side of this PR is that the current `get_data` tests take a couple mins to runs. This of course is not that long of a time, but I think we'd ideally want to reduce the amount of time each test takes as much as possible, as the number of tests will only accumulate over time. Would it be possible to either:

1. limit the spatial/temporal scope of the pytest tests or 
2. create dummy/pre-cached datasets that `get_data` grabs from instead of live retrieving datasets every time the test is run (I bring up 2 as a potential idea to instead create live data tests that wrap around the `retrieve` function rather than `get_data`, since `retrieve` is the common point of retrieving data between `DataParameters` and `get_data`).

vicford commented on climakitae/core/data_interface.py:
```diff
@@ -1693,7 +1705,7 @@ def _list(x):
     ].reset_index(drop=True)
     if len(cat_subset) == 0:
-        print("No data found for your input values")
+        print("ERROR: No data found for your input values")
```
Minor: Can we add "Please modify your data request"

nicolejkeeney commented on climakitae/core/data_interface.py:
```diff
@@ -1664,7 +1672,11 @@ def get_data_options(
-            print("Function arguments require a single string value for your inputs")
+            print(
+                _format_error_print_message(
+                    "Function arguments require a single string value for your inputs"
```
hmm not sure i follow? it seems clear to me, can you suggest an alternative that is more clear?

vicford replied:
From the perspective of a less tech-savvy user, my first thought was "what is a single string". i.e., "single" means what?

nicolejkeeney commented:
> This PR works great for me! Retrieval through `get_data` and `DataParameters` both worked well. I love the addition of the derived indices and making sure `get_data` becomes more and more robust for us all to use.
> 
> One consideration for the test side of this PR is that the current `get_data` tests take a couple mins to run (increasing the total for the test suite to 4min). This of course is not that long of a time, but I think we'd ideally want to reduce the amount of time each test takes as much as possible, as the number of tests will only accumulate over time. Would it be possible to either:
> 
> 1. limit the spatial/temporal scope of the pytest tests or
> 2. create dummy/pre-cached datasets that `get_data` grabs from instead of live retrieving datasets every time the test is run (I bring up 2 as a potential idea because another place to potentially test live retrieval at is at the `retrieve` function directly rather than `get_data`, since `retrieve` is the common point of retrieving data between `DataParameters` and `get_data`).

Good point, I went with option 1 and reduced the temporal scope of the tests. Not sure we want to use dummy data here since the point of the function is to test the data retrieval step. Good idea to keep an eye on the total time elapsed for the tests so it doesn't get out of control at some point; I hadn't thought of that before.
